### PR TITLE
Add support to save models through `tf_save()`

### PR DESCRIPTION
**Pull request by javierluraschi**
Adds support for `tf_save(model, path)`, currently, implemented in `tfestimators` but once the model is proven in `keras`, we would move this signature to `tensorflow`.

The idea behind `tf_save()` is to allow users to export models with ease. This requires each package to provide a `tf_save_spec_from_model()` which will provide a default conversion from a model to a save specification. In most cases, the default specification is enough for proper interoperability. In other cases, explicit specifications would have to be created using `tf_save_regressor_spec()` and `tf_save_classification_spec()`.

A simple example to train and save a model, which could be reused in TensorFlow Serving, follows:

```r
library(tfestimators)
model <- linear_regressor(
 feature_columns = feature_columns( 
    column_numeric("disp", "cyl")
  )
)

mtcars_input_fn <- function(data) {
  input_fn(data, 
           features = c("disp", "cyl"), 
           response = "mpg")
}

model %>% train(mtcars_input_fn(train), steps = 1)
  
tf_save(model, tempdir())
```

### CODE REVIEW

File: R/dnn_estimators.R [hunk 1]
```diff
@@ -69,6 +69,8 @@ dnn_classifier <- function(hidden_units,
                            input_layer_partitioner = NULL,
                            config = NULL)
 {
+  params <- as.list(environment(), all = TRUE)
```
Comment (terrytangyuan) [inline R/dnn_estimators.R]: Could you use some variable name other than `params`? Estimator actually has a `params` field in Python API that we haven't expose in R yet but this i likely to introduce issues.

File: R/save.R [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#' Saves a TensorFlow model with the goal of deploying and reusing across systems.
+#' 
+#' @param model A TensorFlow model. For instance, as generated by \code{linear_regressor()}.
+#' @param path The destination path where the model will be saved.
+#' @param spec The model specification, each model has a default specification;
+#'   but it can be further defined specific functions like \code{tf_save_regressor_spec}.
+#'   
+#' @family save
+#' @export
+tf_save <- function(model, path, spec) {
```
Comment (terrytangyuan) [inline R/save.R]: Maybe something a little more meaningful than `tf_save`, e.g. `tf_save_model`?

File: R/save.R [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  list(
+    feature_columns = model$params$feature_columns,
+    weight_column = model$params$weight_column,
+    label_key = "label"
+  )
+}
+
+#' @export
+#' @family save
+tf_save.tf_estimator_regressor <- function(model, path, spec = tf_save_spec_from_model(model)) {
```
Comment (terrytangyuan) [inline R/save.R]: Regressor -> regression so it's more consistent with tf_estimator_classification?

File: R/save.R [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  UseMethod("tf_save_spec_from_model")
+}
+
+#' @export
+#' @family save
+tf_save_spec_from_model.tf_estimator <- function(model) {
+  feature_columns <- c(
+    model$params$feature_columns,
+    model$params$linear_feature_columns,
+    model$params$dnn_feature_columns
```
Comment (terrytangyuan) [inline R/save.R]: I am not sure why you are extracting all those three types of columns but then this function only needs one of those three - `model$params$feature_columns`. Also other estimators may not have `linear_feature_column` or `dnn_feature_column` so maybe move construct this at specific canned estimator level?

File: R/save.R (hunk 3 above)
Comment (terrytangyuan) [reply R/save.R]: Another issue is that we have to error out like "this is not supported for custom estimators" since for custom estimators you'd have to construct your feature spec like this (as I mentioned in the channel):
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/estimator_test.py#L1475-L1478

**Comment by jjallaire** [comment]
We should discuss where these functions should go as well as naming realtime before we merge. We have to date stayed away from `tf_` prefixing (except for things which are clearly about TF itself like `tf_version` and `tf_config`) because unless we prefix *everything* with `tf_` it's not really clear why one thing is `tf_` and one isn't.

My hope was that for canned estimators we could build support for saving signature defs in so that the user never has to call any special functions to get a servable model. In other words, `export_savedmodel` would by default provide the signature defs for canned estimators, and for custom estimators we'd have helper functions that make constructing the signature defs easier.

I also think that the estimator S3 method definitions should now be moved the **tensorflow** package (e.g. `evaluate` which is already a conflict between tfestimators and keras, as well as `train` and `export_savedmodel`). This will all packages which create flavors of TF model to be able to provide implementations of `export_savedmodel` without having to import tfestimators.

So in short I think this all belongs in the tensorflow package and I'd hope we could get rid of `tf_save` and have that just be `export_savedmodel` (with some helpers to allow easier construction of the serving_input_receiver_fn).

**Comment by javierluraschi** [comment]
@jjallaire right now it's the intention that for canned estimators, users won't need any special helper functions. The rename to `export_savedmodel` and moving this into TF package sounds good to me. Working on this today.

**Comment by jjallaire** [comment]
Note that we already have an `export_savedmodel` function so this will just be an enhancement to that function. I also think that `export_savedmodel` should be an S3 function defined by the tensorflow package so that keras and potentially other packages that build models other ways (e.g. greta) can implement it as well.

Note that it's somewhat controversial to use `export_savedmodel` as the function name because that's already taken by tfestimators. We could just name it `save_model` (and have the S3 method in tfestimators just delegate to `export_savedmodel`.

**Comment by jjallaire** [comment]
We had to force push to master since you originally made this PR. I would start over with a fresh PR onto the new master

**Comment by javierluraschi** [comment]
@jjallaire alright, so I tried to investigate further if we could centralize everything in `tensorflow`; not much success (got close thorugh https://github.com/rstudio/tensorflow/blob/features/export-savedmodel-auto/R/save.R). The challenge is that Tensorflow Serving does expect very specific `tags` while exporting a model, so the save-load functionality is only a subset of what we need on tf-serving.

The good news, by reusing `export_savemodel`, this PR gets simplified since all we are really doing is providing a default function for the default estimators, custom estimators still require input function.

Moving now to investigate and provide a PR for Keras.

**Comment by jjallaire** [comment]
See that the reset against current master has been done. Thanks! Once the @terrytangyuan feedback regarding the name of `params` is addressed I think we are good to merge.

Are there helper functions we can/should export to help custom estimators implement the `serving_input_receiver_fn` or are we best off just pointing them to the `tf$` API?

**Comment by terrytangyuan** [comment]
Looks great! Thank!

File: R/tf_estimator.R [hunk 5]
```diff
@@ -251,12 +254,25 @@ evaluate.tf_estimator <- function(object,
...
                                            assets_extra = NULL,
                                            as_text = FALSE,
                                            checkpoint_path = NULL,
                                            ...)
 {
+  if (is.null(serving_input_receiver_fn)) {
+    if ("tf_custom_estimator" %in% class(object))
+      stop("A 'tf_custom_estimator' requires a custom `serving_input_receiver_fn`.")
+    
+    input_spec <- regressor_parse_example_spec(
```
Comment (terrytangyuan) [inline R/tf_estimator.R]: @javierluraschi Sorry I overlooked this. Seems like here you are only using `regressor_parse_example_spec` but we should check for "regressor" or "classifier" and then choose a parser based on that. Could you fix this as well as adding a test case when you get a chance?

Comment (jjallaire) [reply R/tf_estimator.R]: @javierluraschi Just a reminder to make sure this is somewhere on your todo list (not urgent we do this right away just don't want to lose track of it)
